lib/madwf_transfer.cu

```diff
@@ -0,0 +1,97 @@
+#include <color_spinor_field.h>
+#include <device_vector.h>
+#include <tunable_nd.h>
+#include <madwf_transfer.h>
+#include <kernels/madwf_transfer.cuh>
+
+namespace quda
+{
+  namespace madwf_ml
+  {
+
+    template <class storage_type, class matrix_type> class transfer_5D_wrapper : public TunableKernel3D
+    {
+      ColorSpinorField &out;
+      const ColorSpinorField &in;
+      const matrix_type *wm_p;
+      bool dagger;
+
+    private:
+      unsigned int sharedBytesPerThread() const { return 0; }
+
+      unsigned int sharedBytesPerBlock(const TuneParam &) const { return out.X(4) * in.X(4) * sizeof(matrix_type); }
+
+      unsigned int minThreads() const { return out.VolumeCB() / out.X(4); }
+
+    public:
+      transfer_5D_wrapper(ColorSpinorField &out, const ColorSpinorField &in, const matrix_type *wm_p, bool dagger) :
+        TunableKernel3D(out, out.X(4), out.SiteSubset()), out(out), in(in), wm_p(wm_p), dagger(dagger)
+      {
+        TunableKernel2D_base<false>::resizeStep(out.X(4)); // Ls must be contained in the block
+
+        strcpy(aux, out.AuxString());
+        char tmp[512];
+        sprintf(tmp, ",%02d->%02d", in.X(4), out.X(4));
+        strcat(aux, tmp);
+        strcat(aux, ",transfer_5D");
+        if (dagger) strcat(aux, ",Dagger");
+
+        apply(device::get_default_stream());
+      }
+
+      void apply(const qudaStream_t &stream)
+      {
+        TuneParam tp = tuneLaunch(*this, getTuning(), getVerbosity());
+        if (dagger) {
+          launch<Transfer5D>(tp, stream, Transfer5DArg<storage_type, matrix_type, true>(out, in, wm_p));
+        } else {
+          launch<Transfer5D>(tp, stream, Transfer5DArg<storage_type, matrix_type, false>(out, in, wm_p));
+        }
+      }
+
+      long long flops() const { return 8ll * out.X(4) * 4ll * in.VolumeCB(); }
+      long long bytes() const { return in.Bytes() + out.Bytes(); }
+    };
+
+#ifdef GPU_DOMAIN_WALL_DIRAC
+    template <class transfer_float, transfer_5D_type transfer_type>
+    void transfer_5d_hh(ColorSpinorField &out, const ColorSpinorField &in, const device_vector<float> &tp, bool dagger)
+    {
+      using matrix_type = typename transfer_5D_mapper<transfer_float, transfer_type>::type;
+
+      checkLocation(out, in); // check all locations match
+
+      if (in.SiteSubset() != QUDA_PARITY_SITE_SUBSET || out.SiteSubset() != QUDA_PARITY_SITE_SUBSET) {
+        errorQuda("ColorSpinorFields are not single parity: in = %d, out = %d",
+            in.SiteSubset(), out.SiteSubset());
+      }
+
+      size_t m_size = in.X(4) * out.X(4) * sizeof(matrix_type);
+      if (tp.size() * sizeof(float) != m_size) {
+        errorQuda("Training Parameter size mismatch %lu neq %lu.\n", tp.size() * sizeof(float), m_size);
+      }
+
```

Comment:
Can this be written using `instantiatePreconditioner` or similar to allow us to only compile which precisions are enabled in CMake?

Reply:
Only `float` is instantiated here, and we only need `float`, so I guess there is no need to use something like `instantiatePreconditioner`?

Reply:
https://github.com/lattice/quda/pull/1061/commits/ff335265b55f103ee3927abdf7ef5671034dade2